src/Common/FileCacheFactory.cpp

```diff
@@ -0,0 +1,42 @@
+#include "FileCacheFactory.h"
+#include "FileCache.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+FileCacheFactory & FileCacheFactory::instance()
+{
+    static FileCacheFactory ret;
+    return ret;
+}
+
+FileCachePtr FileCacheFactory::getImpl(const std::string & cache_base_path, std::lock_guard<std::mutex> &)
+{
+    auto it = caches.find(cache_base_path);
+    if (it == caches.end())
+        return nullptr;
+    return it->second;
+}
+
+FileCachePtr FileCacheFactory::getOrCreate(const std::string & cache_base_path, size_t max_size, size_t max_elements_size)
+{
+    std::lock_guard lock(mutex);
+    auto cache = getImpl(cache_base_path, lock);
+    if (cache)
+    {
+        if (cache->capacity() != max_size)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cache with path `{}` already exists, but has different max size", cache_base_path);
```

Comment:
It would be better to just add separate methods `get` and `create`, but the problem is that for the same disk name defined in config, registerDisk for this disk name can be called more than once, therefore I made it `getOrCreate`. But this check for size is a remaining check for user not trying to put the same base path for caches when it was not the intention.